### PR TITLE
improve types of fulfillment helpers

### PR DESCRIPTION
**Pull request by turadg**
## Description

Split from https://github.com/Agoric/agoric-sdk/pull/6963/ for easier review. 

### Security Considerations

--
### Scaling Considerations

--
### Documentation Considerations

--

### Testing Considerations

New tests

### CODE REVIEW

File: packages/inter-protocol/src/collect.js [hunk 1]
```diff
@@ -3,7 +3,7 @@ const { fromEntries, keys, values } = Object;
 /** @type { <X, Y>(xs: X[], ys: Y[]) => [X, Y][]} */
 export const zip = (xs, ys) => harden(xs.map((x, i) => [x, ys[+i]]));
 
-/** @type { <K extends string, V>(obj: Record<K, ERef<V>>) => Promise<Record<K, V>> } */
+/** @type { <T extends Record<string, ERef<any>>>(obj: T) => Promise<{ [K in keyof T]: Awaited<T[K]>}> } */
```
Comment (mhofman) [inline packages/inter-protocol/src/collect.js:6]: 🙏

File: packages/internal/src/utils.js [hunk 2]
```diff
@@ -250,16 +250,12 @@ harden(bindAllMethods);
 
 /**
  * @template {{}} T
- * @typedef {{ [K in keyof T]: DeeplyAwaited<T[K]> }} DeeplyAwaitedObject
+ * @typedef {{ [K in keyof T]: T[K] extends Function ? T[K] : DeeplyAwaited<T[K]> }} DeeplyAwaitedObject
```
Comment (mhofman) [inline packages/internal/src/utils.js]:
```suggestion
 * @typedef {{ [K in keyof T]: T[K] extends (...args: any[]) => any ? T[K] : DeeplyAwaited<T[K]> }} DeeplyAwaitedObject
```

File: packages/internal/src/utils.js [hunk 3]
```diff
@@ -250,16 +250,12 @@ harden(bindAllMethods);
...
  */
 
 /**
- * Caveats:
- * - doesn't recur within Promise results
- * - resulting type has wrapper in its name
- *
  * @template T
- * @typedef {T extends PromiseLike<any> ? Awaited<T> : T extends {} ? DeeplyAwaitedObject<T> : Awaited<T>} DeeplyAwaited
+ * @typedef {T extends PromiseLike<any> ? Awaited<T> : T extends {} ? import('type-fest').Simplify<DeeplyAwaitedObject<T>> : Awaited<T>} DeeplyAwaited
```
Comment (mhofman) [inline packages/internal/src/utils.js]: Is this the first time we take an external dependency for our types? Wondering how that works out once we start publishing our packages correctly and the exported types have external dependencies (there is an argument to be made in that case that it is not in fact a dev dependency)

**Review by mhofman — COMMENTED** [review]
Looks good but I'll hold approval until clarification on the external dependency, and the fix for extends callable vs Function

File: packages/internal/src/utils.js (hunk 2 above)
Comment (turadg) [reply packages/internal/src/utils.js]: Done with a`Callable` type

File: packages/internal/src/utils.js (hunk 3 above)
Comment (turadg) [reply packages/internal/src/utils.js]: Good catch! `yarn prepack` showed this change didn't even build.

It's such a small type I backed out the import and copied it over with attribution.

File: packages/internal/src/utils.js [hunk 4]
```diff
@@ -248,18 +248,25 @@ export const bindAllMethods = obj =>
   );
 harden(bindAllMethods);
 
+/**
+ * @template T
+ * @typedef {{[KeyType in keyof T]: T[KeyType]} & {}} Simplify
+ * flatten the type output to improve type hints shown in editors
```
Comment (mhofman) [inline packages/internal/src/utils.js:254]: Neat

**Review by mhofman — APPROVED** [review]
Thanks for these improvements!

**Pull request merged**